### PR TITLE
 Vast Tracker Injection Design Review [DRAFT]

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
This PR consist of different approaches for vast event tracker injection.

Issue: https://github.com/prebid/prebid-server/issues/1725

BenchMark:

```
// BenchmarkXMLParser-12    	        10081	         114797 ns/op        38260 B/op	     333 allocs/op
// BenchmarkXMLEncoding-12              10802	         109733 ns/op        38601 B/op	     286 allocs/op
// BenchmarkStructUnmarshal-12          4402	         239360 ns/op        72024 B/op	     829 allocs/op
// BenchmarkStringParser-12    	        34964	         31871  ns/op        28325 B/op	     30 allocs/op
// BenchmarkETree-12    	        5698	         180206 ns/op        61769 B/op	     792 allocs/op
```

### CODE REVIEW

File: exchange/injector/vaststruct/injector.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+type Node struct {
+	XMLName xml.Name
+	Attrs   []xml.Attr `xml:",any,attr"`
+	Content []byte     `xml:",cdata"`
+	// Nodes   []Node     `xml:",any"`
+}
+
+func main() {
```
Comment (AlexBVolcy) [inline exchange/injector/vaststruct/injector.go:147]: This main function needs code coverage

File: exchange/injector/xml_encoder/injector.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+							outputXML.WriteString(typ)
+							outputXML.WriteString("\"><![CDATA[")
+							builder.replacer.ReplaceBytes(&outputXML, url, builder.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					outputXML.WriteString("</TrackingEvents>")
+					encoder.EncodeToken(tt)
+				}
+			case "Linear":
```
Comment (AlexBVolcy) [inline exchange/injector/xml_encoder/injector.go:121]: This linear case needs code coverage

**Comment by SyntaxNode** [comment]
@AlexBVolcy I think you may have misunderstood the intent of this PR. It proposes several algorithms for implementing XML injector tracking. @Pubmatic-Dhruv-Sonone is seeking our input on which approach is preferred.

**Comment by SyntaxNode** [comment]
@Pubmatic-Dhruv-Sonone Thank you for implementing on all of the options and including detailed benchmark results.

While StringParser is the clear winner in terms of speed and allocations, I'm worried how it might behave in the case of invalid xml, which may be intentional as an attack on the system. We've had issues a long time ago with string shortcuts for json generation. What do others think about this?

XMLParser and XMLEncoder are essentially tied for first among the options which validate the xml structure. The Encoder has a slight memory advantage with cleaner code IMHO, making it a winner in my view. Rather than constantly flushing the encoder and outputting xml directly to the output buffer, could use the encoder to construct those tags?

**Comment by Pubmatic-Dhruv-Sonone** [comment]
> @Pubmatic-Dhruv-Sonone Thank you for implementing on all of the options and including detailed benchmark results.
> 
> While StringParser is the clear winner in terms of speed and allocations, I'm worried how it might behave in the case of invalid xml, which may be intentional as an attack on the system. We've had issues a long time ago with string shortcuts for json generation. What do others think about this?
> 
> XMLParser and XMLEncoder are essentially tied for first among the options which validate the xml structure. The Encoder has a slight memory advantage with cleaner code IMHO, making it a winner in my view. Rather than constantly flushing the encoder and outputting xml directly to the output buffer, could use the encoder to construct those tags?

@SyntaxNode  Updated the XMLEncoder approach to encode tags instead of flushing to output buffer.
Below is the benchmark for the updated approach. The  performance has degraded after the changes.
```
// goos: darwin
// goarch: amd64
// pkg: github.com/prebid/prebid-server/exchange/injector/xml_encoder
// cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
// BenchmarkXMLEncoding-12    	    6666	    158149 ns/op	   53132 B/op	     398 allocs/op
// PASS
// ok  	github.com/prebid/prebid-server/exchange/injector/xml_encoder	1.395s
```

Also in XMLEncoder approach, sequence of the tags  is not  maintained as we are injecting child tag just before the end of  parent tags. 

Ex If vast xml already has a **Impression** tag then this approach injects new Impression tag before the ending **Inline** or **Wrapper** tag.  This might break the vast xsd validation as it expected similar tags to be grouped together.

**Comment by Pubmatic-Dhruv-Sonone** [comment]
@SyntaxNode  I changed XMLEncoder approach as suggested by you, Do you have anymore thoughts regarding the approaches?

**Comment by SyntaxNode** [comment]
@Pubmatic-Dhruv-Sonone Since I think we've selected a design, can we close this PR? You can open a new PR with the final code when you're ready.

**Pull request closed without merging**